src/SFA.DAS.EmployerFinance/Application/Commands/ProcessLevyDeclarations/ProcessLevyDeclarationsCommandHandler.cs

```diff
@@ -0,0 +1,27 @@
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+using SFA.DAS.EmployerFinance.Data;
+using SFA.DAS.EmployerFinance.Models;
+
+namespace SFA.DAS.EmployerFinance.Application.Commands.ProcessLevyDeclarations
+{
+    public class ProcessLevyDeclarationsCommandHandler : AsyncRequestHandler<ProcessLevyDeclarationsCommand>
+    {
+        private readonly EmployerFinanceDbContext _db;
+
+        public ProcessLevyDeclarationsCommandHandler(EmployerFinanceDbContext db)
+        {
+            _db = db;
+        }
+        
+        protected override async Task Handle(ProcessLevyDeclarationsCommand request, CancellationToken cancellationToken)
+        {
+            var accountPayeSchemes = await _db.AccountPayeSchemes.AsNoTracking().ToListAsync(cancellationToken);
+            var saga = new LevyDeclarationSaga(request.PayrollPeriod, accountPayeSchemes);
```

Comment:
There _might_ be a race condition here (but without serious consequences 🙂).

`LevyDeclarationSaga`'s constructor publishes a `StartedProcessingLevyDeclarationsEvent`, which among other things publishes an `ImportLevyDeclarationsCommand` whose handler assumes the `LevyDeclarationSaga` is in the database. The `LevyDeclarationSaga` is saved in this handler.

Whether this is a race condition or not depends on the order in which the `NServiceBus` and `EntityFrameworkCore` `UnitOfWork`'s commit (once they're added), which I'm unsure of. Even if there is a race condition, it should only cause the `StartedProcessingLevyDeclarationsEvent` message handler to initially error (I think), but subsequent retries should work once the saga is in the database.

Reply:
You're correct, the data persistence and message publishing will happen within the same transaction.